### ISSUE TITLE
Output as Text

### ISSUE THREAD

**Issue by mike-ward** [Knagis/CommonMark.NET#26]
I'm back with my hat in hand :)

Plain text output please, with hard line breaks at a set column.

Currently I do this with pandoc

    pandoc -f markdown -t markdown --columns=80 document.md > newdocument.md

Then I could build refomatter into Markdown Edit.

**Comment by Knagis** [comment]
To clarify - you are not looking for plain text output but rather markdown output, preserving all the little details?

One problem that I see right away - inline links and reference links - the details on how they were entered are not preserved. So the formatter has to choose the resulting formatting - but what should it be? For example, are the references added after the paragraph or at the end? Does it use numeric reference labels or creates some magic strings?

**Comment by mike-ward** [comment]
Correct, I want to preservie the mMarkdown although some interpretation, as you point out, will occur. I don't use references so I can't speak to the issue. Pandoc reformats # headers into === and --- headers. I imagine there are some other things it does like thatl, but nothing I have ever been bothered by. Hard line breaks make reading text from the console a bit nicer.

**Comment by Knagis** [comment]
Header texts can be left alone - AST contains information from which format it was parsed. I would avoid converting them between formats.

For now reference links seem to be the only thing that loses significant information during parsing.

**Comment by Knagis** [comment]
I have started work on this in the [md-writer](https://github.com/Knagis/CommonMark.NET/tree/md-writer) branch. Perhaps you could try it out and list the issues you can find.

Some I know already:
- [ ] hard-wrapping lines are not yet implemented
- [ ] blockquote prefix `> ` is not rendered on blank lines within the quote
- [ ] links and images loses title attribute

**Comment by mike-ward** [comment]
ooooh, new shiny to play with this weekend. Thanks!

**Comment by mike-ward** [comment]
Hey, this is looking nice. Here's something I noticed

    Header Text
    =========

is formatted as

    Header Text
    ===

I think in the spirit of, "keeping it pretty", we should either keep the number of underlines elements in the original document or better still, match the number of underline elements to the length of the text.

The parser correctly  preserves `# Header` elements. +1

In ordered and unordered lists, the original indent is not preserved

      1. line 1
      2. line 2

is rendered as

    1. line 1
    2. line 2

This is not big issue. However, when I write markdown I usually indent these items a couple of spaces to simulate how they appear in HTML

If preserving the original indentation is difficult, I would settle for a configuration option. 

Same thing goes for blocks quotes. Source indentation in not preserved.

Finally, the following:

      1. hello
      2. there
    
      > block
      > quote

renders as:

    1. hello
    2. there
    > block quote

Should there be a blank line b/t the numbered list and block quote?

That's all for now. Great job!

**Comment by mike-ward** [comment]
Horizontal breaks present some interesting challenges.

    ----------------

is  currently output as

    ---

Couple of options here.

1. Keep current behavior. It's correct but not stylish
2. Perserve the original number of dashes
3. Pad with dashes to the column length

Not sure what I want here. Probably option 3 but 2 would likey suffice.

**Comment by Knagis** [comment]
> header texts

I will change this to match the number of characters in the previous line

> In ordered and unordered lists, the original indent is not preserved

I think that the parser preserves this information so the list could be indented based on the first item.

> block quote after list

that is a bug coming from the fact that the list is tight so the paragraphs do not render the newline. this should be easy to fix.

> horizontal breaks

I suppose that the only reasonable solution is to add a configuration option on how many dashes to add. Leaving the count from the source goes against the idea of reformatting the document accoring to a single style.

Thank you for the feedback!

**Comment by mike-ward** [comment]
I agree on all your points.

**Comment by mike-ward** [comment]
One other thought. You might want to include a configuration to always force H1/H2 headers to underline. Or if you're really crazy a 3 way setting.

1. always underlined
2. always #
3. keep document format

**Comment by mike-ward** [comment]
I'm looking for a new feature to add to Markdown Edit this weekend. Hint, hint, hint :smile:

**Comment by Knagis** [comment]
Unfortunately I had no time to work on this so it is still in the state you last reviewed...

**Comment by kieranbenton** [comment]
Any movement on this guys? Is it still living in the branch? Any idea what kind of work is outstanding and if I can contribute to get it out and into a release? Cheers.

**Comment by jazzdelightsme** [comment]
Just FYI, I am interested in a markdown parser that can do full-fidelity "round trip", as well as normalization on top of that. The work described in this thread seems aimed at normalization, but being able to just do round-trip (emitting the exact same text as the original output) is also useful when you want to do programmatic transformations without mucking about with the existing style.
